consensus/expected.go

```diff
@@ -278,29 +283,47 @@ func (c *Expected) RunStateTransition(ctx context.Context, ts types.TipSet, tsMe
 //    Returns nil if all the above checks pass.
 // See https://github.com/filecoin-project/specs/blob/master/mining.md#chain-validation
 func (c *Expected) validateMining(ctx context.Context, st state.Tree, ts types.TipSet, parentTs types.TipSet) error {
+
+	prevTicket, err := parentTs.MinTicket()
+	if err != nil {
+		return errors.Wrap(err, "failed to read parent min ticket")
+	}
+	prevHeight, err := parentTs.Height()
+	if err != nil {
+		return errors.Wrap(err, "failed to read parent height")
+	}
 	for i := 0; i < ts.Len(); i++ {
 		blk := ts.At(i)
-		// TODO: Also need to validate BlockSig
-
-		// TODO: Once we've picked a delay function (see #2119), we need to
-		// verify its proof here. The proof will likely be written to a field on
-		// the mined block.
 
-		// See https://github.com/filecoin-project/specs/blob/master/mining.md#ticket-checking
-		result, err := IsWinningTicket(ctx, c.bstore, c.PwrTableView, st, blk.Tickets[0], blk.Miner)
+		workerAddr, err := c.PwrTableView.WorkerAddr(ctx, st, c.bstore, blk.Miner)
 		if err != nil {
-			return errors.Wrap(err, "can't check for winning ticket")
+			return errors.Wrap(err, "error reading worker address of block miner")
+		}
+		// Validate ElectionProof
+		result, err := c.IsElectionWinner(ctx, c.bstore, c.PwrTableView, st, blk.Tickets[0], blk.ElectionProof, workerAddr, blk.Miner)
+		if err != nil {
+			return errors.Wrap(err, "failed checking election proof")
 		}
-
 		if !result {
-			return errors.New("not a winning ticket")
+			return errors.New("block author did not win election")
 		}
-	}
-	return nil
-}
 
+		// Validate ticket array
+		// TODO #2223 allow multiticket arrays and check ticket array
+		// length against nullblock count
+		if len(blk.Tickets) != 1 {
+			return errors.New("invalid ticket array length")
+		}
+		result, err = c.ValidateTicket(prevTicket, blk.Tickets[0], workerAddr, uint64(blk.Height)-prevHeight-1)
+		if err != nil {
+			return errors.Wrap(err, "failed checking ticket array")
+		}
+		if !result {
+			return errors.New("invalid ticket extension")
+		}
 
 	}
+	return nil
```

Comment:
Commentary: this looks like stuff that should move to the BlockSemanticValidator soon.